pinot-common/src/main/java/com/linkedin/pinot/common/utils/MmapUtils.java

```diff
@@ -58,16 +59,71 @@
   }
 
   public static class AllocationContext {
+    private final String fileName;
+    private final String parentFileName;
+    private final String parentPathName;
+
     private final String context;
     private final AllocationType allocationType;
 
-    public AllocationContext(String context, AllocationType allocationType) {
+    public static final Joiner PATH_JOINER = Joiner.on(File.separatorChar).skipNulls();
+
+    AllocationContext(String context, AllocationType allocationType) {
       this.context = context;
       this.allocationType = allocationType;
+      this.fileName = null;
+      this.parentFileName = null;
+      this.parentPathName = null;
+    }
+
+    AllocationContext(File file, String details, AllocationType allocationType) {
+      context = details;
+      this.allocationType = allocationType;
+
+      // We separate the path into three components to lower the overall on-heap memory usage when there are lots of
+      // redundant paths. Paths that are memory mapped are usually <storage directory>/<segment name>/<column name>.ext
+      // and this allows sharing the same String instance if the storage directory and segment name is duplicated many
+      // times.
+
+      if (file != null) {
```

Comment:
is it possible to write a regex to extract out the directory, segmentment name, columnName.ext ?